### ISSUE TITLE
[FSDP] mixed precision buffers are only maintained in FP16

### ISSUE THREAD

**Issue by myleott** [facebookresearch/fairscale#442]
## 🐛 Bug

In FSDP with `mixed_precision=True`, we only maintain buffers in FP16. This could lead to underflow/overflow for buffers that wouldn't happen with standard AMP training.

Ideally we could keep the buffers in their original dtype, but it's unclear how/when/if to cast to `compute_dtype`. For static buffers it doesn't matter, but for BatchNorm we want updates to happen in full precision.

**Comment by sshleifer** [comment]
I have struggled to create a regression test where we get over/underflow unless buffers get put back into fp32 after `fwd`. 
The problem is that the dynamic buffer I tried, `torch.nn.BatchNorm1d().running_mean` gets updating during forward, so we have no overflow.  I think this is part of the definition of buffer (no `.grad`, so no updates during `optim.step()`, so the only reason for casting buffers out of `compute_dtype` is (a) state_dict and (b) maintaining the fp32 after forward invariant. 

Am I misunderstanding the problem @min-xu-ai @myleott ?

**Comment by min-xu-ai** [comment]
Is over/underflow absolutely needed in the unit test or can we assert it is equal to non-FSDP training (DDP + AMP?) being sufficient?

> Ideally we could keep the buffers in their original dtype, but it's unclear how/when/if to cast to compute_dtype. For static buffers it doesn't matter, but for BatchNorm we want updates to happen in full precision.

For the statement above, I am don't quite understand "static buffers doesn't matter part". Also, BN.half() seems like a method users are supposed to call when they do mixed precision. Should we call it too?

**Comment by myleott** [comment]
If using PyTorch amp, BatchNorm should stay/happen in FP32:
```python
import torch
from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
from fairscale.utils.testing import DummyProcessGroup

bn = torch.nn.BatchNorm1d(10).cuda()

with torch.cuda.amp.autocast():
    input = torch.randn(20, 10).cuda()
    output = bn(input)
print(output.dtype)  # torch.float32

group = DummyProcessGroup(rank=0, size=1)
fsdp_bn = FSDP(bn, mixed_precision=True, process_group=group)
with torch.cuda.amp.autocast():
    input = torch.randn(20, 10).cuda()
    output = fsdp_bn(input)
print(output.dtype)  # torch.float16
```

**Comment by sshleifer** [comment]
Extended your example to contain a static buffer ("`float_buffer`") created by `register_buffer` and a `BatchNorm1d ` and test amp and apex O2 behavior:
 
- Autocast: leaves `float_buffer` and `bn.running_mean` in fp32.
- Apex `O2` treats `float_buffer` and `bn.running_mean` differently. Both return tensors in `16` but `bn.running_mean.dtype` is `32`, whereas `float_buffer.dtype` is `16`.

Since positional embeddings in NLP are usually at the beginning of the model, `autocast` behavior seems undesirable, so maybe the **desired** behavior is

- `float_buffer`: `16` during `forward`, `32` after forward.
- `bn`: `O2` magic during forward, `32` after forward.

This contradicts Myle's comment in #449, "I think we want `buffer_dtype` to stay fp32 inside the forward pass even when using mixed precision", which seems like an easier thing to implement, so I wanted to check my understanding.
 
### Script

```python
import torch
from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
from fairscale.utils.testing import DummyProcessGroup
from torch import nn


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.bn = torch.nn.BatchNorm1d(10)
        self.l1 = nn.Linear(10, 10)
        self.register_buffer('float_buffer', self.l1.weight.data[0] / 2.5)
    def forward(self, x):
        print(f'input: {x.dtype}')
        x = self.l1(x)
        print(f'after l1: {x.dtype}')
        x = x / self.float_buffer
        print(f'after float buffer: {x.dtype}')
        x = self.bn(x)
        print(f'bn output: {x.dtype}')
        print(f'bn.running_mean: {self.bn.running_mean.dtype}')
        print(f'bn.weight: {self.bn.weight.dtype}')
        print(f'float_buffer: {self.float_buffer.dtype}')
        return x

model = Model().cuda()
print('Autocast:')
with torch.cuda.amp.autocast():
    input = torch.randn(20, 10).cuda()
    output = model(input)

import apex

model = Model().cuda()
optimizer = torch.optim.SGD(model.parameters(), lr=.01, momentum=0.9)
print('')
print('Apex O2:')
model, optimizer = apex.amp.initialize(model, optimizer, opt_level="O2", verbosity=0)
input = torch.randn(20, 10).cuda()
x = model(input)
```

### Output 
```python
Autocast:
input: torch.float32
after l1: torch.float16
after float buffer: torch.float32
bn output: torch.float32
bn.running_mean: torch.float32
bn.weight: torch.float32
float_buffer: torch.float32

Apex O2:
input: torch.float16
after l1: torch.float16
after float buffer: torch.float16
bn output: torch.float16
bn.running_mean: torch.float32
bn.weight: torch.float32
float_buffer: torch.float16
```

**Comment by myleott** [comment]
New script looks great!

You're right that it looks like Apex O2 is doing some magic, probably patching the batch norm op specifically :/

Autocast is a bit simpler. There's a [list of ops](https://pytorch.org/docs/master/amp.html#op-specific-behavior) that automatically cast from fp32 -> fp16 and a similar list that automatically cast from fp16 -> fp32. All other ops happen in the original dtype. But it seems that BatchNorm1d isn't on the fp16 -> fp32 list, even though by all accounts it should be done in fp32 to preserve convergence in vision models. See this discussion: https://discuss.pytorch.org/t/autocast-with-normalization-layers/94125/5

> Since positional embeddings in NLP are usually at the beginning of the model, autocast behavior seems undesirable

Is your worry that the positional embed lookup will be FP32 and then all ops after will get stuck there? This should only affect the initial `embed_tokens(ids) + embed_positions(ids)` op, since on the first linear op everything should switch back to fp16.

Anyway, this is all pretty confusing, and it's not clear to me there's a single preferred solution. What do you think about exposing a `buffer_dtype` config option?

> `float_buffer`: `16` during `forward`, `32` after forward.
> `bn`: `O2` magic during forward, `32` after forward.

I'm wondering if we even want 32 after forward... it seems like buffers should only stay in one dtype, since there's no downstream dependency on an optimizer, we're not actually maintaining two copies, etc. So maybe we just cast once to a given dtype and leave buffers in that dtype forever?

**Comment by sshleifer** [comment]
> Is your worry that the positional embed lookup will be FP32 and then all ops after will get stuck there? This should only affect the initial embed_tokens(ids) + embed_positions(ids) op, since after the first Linear op everything should switch back to fp16.

Yes, I was confused, thanks!

>What do you think about exposing a `buffer_dtype` config option?

Yes that's very easy and seems to allow full control. Implemented in #458.
